File: problems/3194_Minimum_Average_of_Smallest_and_Largest_Elements/solution.ts
export {}

/**
 * NOTE: If it wasn't asked specifically to be solved like this,
 * it would be more efficient to not push the averages, rather
 * compare the the current average to the existing min and update
 * that one instead.
 */
function minimumAverage(nums: number[]): number {
  const averages = []

  const sortedNums = nums.sort((a, b) => b - a)

  while (sortedNums.length) {
    const min = sortedNums.shift()!
    const max = sortedNums.pop()!

    averages.push((min + max) / 2)
  }

  return Math.min(...averages)
}

/**
 * Time complexity: O(log n + n / 2) => O(n)
 * Space complexity: O(n)
 */

console.log(minimumAverage([7, 8, 3, 4, 15, 13, 4, 1]))
console.log(minimumAverage([1, 9, 8, 3, 10, 5]))
console.log(minimumAverage([1, 2, 3, 7, 8, 9]))

File: problems/3194_Minimum_Average_of_Smallest_and_Largest_Elements/solution.go
package main

import (
	"fmt"
	"math"
	"sort"
);

func minimumAverage(nums []int) float64 {
	averages := []float64{}

	sortedNums := nums;
	sort.Ints(sortedNums);

	for len(sortedNums) > 0 {
		min := sortedNums[0]
		sortedNums = sortedNums[1:]
		max := sortedNums[len(sortedNums) - 1]
		sortedNums = sortedNums[:len(sortedNums) - 1]

		averages = append(averages, float64(min + max) / 2)
	}

	min := averages[0]

	for _, val := range averages[1:] {
        min = math.Min(min, val)
    }

	return min
}

/**
 * Time complexity: O(log n + n / 2) => O(n)
 * Space complexity: O(n)
 */

func main() {
	fmt.Println(minimumAverage([]int{7, 8, 3, 4, 15, 13, 4, 1}));
	fmt.Println(minimumAverage([]int{1, 9, 8, 3, 10, 5}));
	fmt.Println(minimumAverage([]int{1, 2, 3, 7, 8, 9}));
}

